File: 01-week-1--starter-algorithms/03-day-4--selection-sort/javascript/selection_sort.js
function selectionSort(arr) {
  // type your code here
  // let x
  let leng = arr.length
  for (let i=0; i < leng; i++){
      let y = i
    for(let x=i+1; x < leng; x++){
      if (arr[x] < arr[y]) {
        y = x
      }
    }
    if (y != i) {
      [arr[y], arr[i]] = [arr[i], arr[y]]
    }
  }
    return arr

}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSort([3, -1, 5, 2]));

  console.log("");

  // BENCHMARK HERE, and print the average runtime
  const longInput = [];

  for (let i = 0; i < 100; ++i) {
    longInput.push(Math.random());
  }
}

module.exports = selectionSort;

// Please add your pseudocode to this file
// create a temprorary variable for the array length so we keep a scope a level deeper
// for each element in the array
// create a temporary variable to store the original value
// for each subsequent element in the array
// if the latter value is less than the former
// set that latter value in the place of the former
// if the new value at that index is different than the stored original value
// swap them
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/03-day-4--selection-sort/ruby/selection_sort.rb
def selection_sort(arr)
  # type your code in here
  (arr.length - 1).times do |element_index_to_check|
    current_min_index = element_index_to_check + 1
    element_index_to_check.upto(arr.length - 1) do |left_num_index|
      if arr[left_num_index] < arr[current_min_index]
        current_min_index = left_num_index
      end
    end
    temp_left_num = arr[current_min_index]
    arr[current_min_index] = arr[element_index_to_check]
    arr[element_index_to_check] = temp_left_num
  end
    arr
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort([3, -1, 5, 2])

  puts

  # Don't forget to add your own!

  # BENCHMARK HERE, you can print the average runtime
  long_input = []

  100.times { long_input << rand }
end

# Please add your pseudocode to this file
# arr length - 1 ignores zero. this is how many times we have to run this iteration
# assume the second number is the minimum and store it as a temporary variable
# check from the 0 index and work to the right
# if the latter number is less than the former number
# set that latter number as the current minimum between the two
# store the value of the second element
# replace it with the new smallest
# And a written explanation of your solution
